In [22]:
import sklearn
import pandas as pd
import numpy as np
from sklearn.cross_validation import train_test_split

In [3]:
df_wine = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/wine/wine.data", header=None)

In [4]:
df_wine.head()

0      1     2     3     4    5     6     7     8     9     10    11    12  \
0   1  14.23  1.71  2.43  15.6  127  2.80  3.06  0.28  2.29  5.64  1.04  3.92   
1   1  13.20  1.78  2.14  11.2  100  2.65  2.76  0.26  1.28  4.38  1.05  3.40   
2   1  13.16  2.36  2.67  18.6  101  2.80  3.24  0.30  2.81  5.68  1.03  3.17   
3   1  14.37  1.95  2.50  16.8  113  3.85  3.49  0.24  2.18  7.80  0.86  3.45   
4   1  13.24  2.59  2.87  21.0  118  2.80  2.69  0.39  1.82  4.32  1.04  2.93   

     13  
0  1065  
1  1050  
2  1185  
3  1480  
4   735

In [7]:
df_wine.columns = ['Class Label', 'Alcohol', 'Malic Acid', 'Ash', 'Alcalinity of Ash', 'Magnesium',
                   'Total phenols', 'Flavanoids', 'Nonflavinoid phenols', 'Proanthocyanins', 'Color intensity', 'Hue', 'OD280/OD315 of diluted wines', 
                   'Proline']

In [19]:

df_wine['Alcohol'] = df_wine['Alcohol'] / 10
df_wine['Alcalinity of Ash'] = df_wine['Alcalinity of Ash'] / 10
df_wine['Magnesium'] = df_wine['Magnesium'] / 100
df_wine['Proline'] = df_wine['Proline'] / 1000
df_wine.head()

Class Label  Alcohol  Malic Acid   Ash  Alcalinity of Ash  Magnesium  \
0            1    1.423        1.71  2.43               1.56       1.27   
1            1    1.320        1.78  2.14               1.12       1.00   
2            1    1.316        2.36  2.67               1.86       1.01   
3            1    1.437        1.95  2.50               1.68       1.13   
4            1    1.324        2.59  2.87               2.10       1.18   

   Total phenols  Flavanoids  Nonflavinoid phenols  Proanthocyanins  \
0           2.80        3.06                  0.28             2.29   
1           2.65        2.76                  0.26             1.28   
2           2.80        3.24                  0.30             2.81   
3           3.85        3.49                  0.24             2.18   
4           2.80        2.69                  0.39             1.82   

   Color intensity   Hue  OD280/OD315 of diluted wines  Proline  
0             5.64  1.04                          3.92    1.065  
1             4.38  1.05                          3.40    1.050  
2             5.68  1.03                          3.17    1.185  
3             7.80  0.86                          3.45    1.480  
4             4.32  1.04                          2.93    0.735

In [20]:
X = df_wine.iloc[:, 1:].values
y = df_wine.iloc[:, 0].values

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

In [23]:
from sklearn.svm import SVC
svm = SVC(kernel='linear', C=1.0, random_state=0)
svm.fit(X_train, y_train)

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape='ovr', degree=3, gamma='auto', kernel='linear',
  max_iter=-1, probability=False, random_state=0, shrinking=True,
  tol=0.001, verbose=False)

In [24]:
y_pred = svm.predict(X_test)

In [25]:
def accuracy_report(y_pred, y_actual):
    val = [(y_pred[i] == y_actual[i]) for i in range(len(y_pred))]
    return sum(val) / len(val)

In [26]:
accuracy_report(y_pred, y_test)

0.9444444444444444

In [27]:
from sklearn.ensemble import RandomForestClassifier
forest = RandomForestClassifier(criterion='entropy', n_estimators=100, random_state=1, n_jobs=2)

In [28]:
forest.fit(X_train, y_train)

RandomForestClassifier(bootstrap=True, class_weight=None, criterion='entropy',
            max_depth=None, max_features='auto', max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=1, min_samples_split=2,
            min_weight_fraction_leaf=0.0, n_estimators=100, n_jobs=2,
            oob_score=False, random_state=1, verbose=0, warm_start=False)

In [29]:
yh = forest.predict(X_test)

In [31]:
accuracy_report(yh, y_test)

0.9629629629629629

# Results:
- so a random forest with 100 trees gets 96.3%
- a SVM with a linear kernel gets 94.4 % 

In [41]:
from sklearn.neighbors import KNeighborsClassifier
knn = KNeighborsClassifier(n_neighbors=3, p=2, metric='minkowski') #uses Euclidian because thats what minkowski is at p=2

In [42]:
knn.fit(X_train, y_train)

KNeighborsClassifier(algorithm='auto', leaf_size=30, metric='minkowski',
           metric_params=None, n_jobs=1, n_neighbors=3, p=2,
           weights='uniform')

In [43]:
yp = knn.predict(X_test)

In [44]:
accuracy_report(yp, y_test)

0.9074074074074074

In [45]:
#three nearest neighbors is better than 5? why? because less overfitting ! 